### PR TITLE
Fix LR scheduler behaviour with AMP

### LINKED ISSUE #16228: Wrong LR scheduler behaviour when using AMP
### Bug description

When training with native AMP and a LR scheduler, we get a warning that indicates that a LR step has been taken when an optimizer step was skipped (expected at the beginning of the training with native AMP):

This can be fixed by wrapping these lines https://github.com/Lightning-AI/lightning/blob/574a9516012b4ab778254055c537f5d57e8e694f/src/pytorch_lightning/core/module.py#L1589-L1592

in `if hasattr(optimizer, '_step_count') and optimizer._step_count > 0`.

### How to reproduce the bug

```python
import os

import torch
from torch.utils.data import DataLoader, Dataset

from pytorch_lightning import LightningModule, Trainer


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum() * 100000000
        self.log("train_loss", loss)
        return {"loss": loss}

    def validation_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("valid_loss", loss)

    def test_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("test_loss", loss)

    def configure_optimizers(self):
        opt = torch.optim.SGD(self.layer.parameters(), lr=0.1)
        sched = torch.optim.lr_scheduler.StepLR(opt, 10)
        return {"optimizer": opt, 'lr_scheduler': {"scheduler": sched, "interval": "step"}}

def run():
    train_data = DataLoader(RandomDataset(32, 32), batch_size=2)

    model = BoringModel()
    trainer = Trainer(
        accelerator='gpu',
        devices=1,
        default_root_dir=os.getcwd(),
        num_sanity_val_steps=0,
        max_epochs=1,
        precision=16,
        enable_model_summary=False,
    )
    trainer.fit(model, train_dataloaders=train_data)


if __name__ == "__main__":
    run()
```

### Error messages and logs

```
/usr/local/lib/python3.8/dist-packages/torch/optim/lr_scheduler.py:138: UserWarning: Detected call of `lr_scheduler.step()` before `optimizer.step()`. In PyTorch 1.1.0 and later, you should call them in the opposite order: `optimizer.step()` before `lr_scheduler.step()`.  Failure to do this will result in PyTorch skipping the first value of the learning rate schedule. See more details at https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate
```

### Environment

_No response_

### More info

_No response_

### PR DESCRIPTION
**Pull request by milesial**
## What does this PR do?

When training when native AMP and a LR scheduler, we get this warning that indicates that a LR step has been taken when an optimizer step was skipped (expected at the beginning of the training with native AMP):
```
/usr/local/lib/python3.8/dist-packages/torch/optim/lr_scheduler.py:138: UserWarning: Detected call of `lr_scheduler.step()` before `optimizer.step()`. In PyTorch 1.1.0 and later, you should call them in the opposite order: `optimizer.step()` before `lr_scheduler.step()`.  Failure to do this will result in PyTorch skipping the first value of the learning rate schedule. See more details at https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate
```

Fixes #16228 #5558

### Does your PR introduce any breaking changes? If yes, please list them.

No

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

File: src/pytorch_lightning/core/module.py [hunk 1]
```diff
@@ -1585,6 +1585,10 @@ def lr_scheduler_step(self, scheduler, optimizer_idx, metric):
                 scheduler.step(epoch=self.current_epoch)
 
         """
+        optimizer = self.trainer.optimizers[optimizer_idx]
+        if hasattr(optimizer, "_step_count") and optimizer._step_count <= 0:
```
Comment (carmocca) [inline src/pytorch_lightning/core/module.py:1589]: This only works for the first step but the warning can appear after any optimizer step, as there's no easy way to know whether the `optimizer.step` was skipped by AMP.

Comment (milesial) [reply src/pytorch_lightning/core/module.py:1589]: This works for the N first "steps" that are skipped. True that we can't detect skipped steps passed the first true step, but PyT doesn't print any warning for those https://github.com/pytorch/pytorch/blob/master/torch/optim/lr_scheduler.py#L138

**Comment by milesial** [comment]
In the process of fixing tests I discovered and fixed a bug where the scheduler wouldn't match its optimizer when multiple optimizers are instantiated with frequencies. Now the optimizers and schedulers match and alternate as they should, resetting the cycle every epoch.

File: src/pytorch_lightning/core/module.py (hunk 1 above)
Comment (akihironitta) [reply src/pytorch_lightning/core/module.py:1589]: > True that we can't easily detect skipped steps passed the first true step, but PyT doesn't print any warning for those

No, but you can detect it by comparing the scale factor before and after the `scaler.step()` as I wrote in the linked issue. IMHO, I wouldn't see the current change as a solution for the reason @carmocca stated.

Comment (milesial) [reply src/pytorch_lightning/core/module.py:1589]: Ok, I can change the implementation to a comparison of scales if that's prefered

Comment (milesial) [reply src/pytorch_lightning/core/module.py]: Done

**Comment by milesial** [comment]
@carmocca Ready for final review

File: src/pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 2]
```diff
@@ -390,7 +391,7 @@ def update_lr_schedulers(self, interval: str, update_plateau_schedulers: bool) -
         if interval == "step" and self._should_accumulate():
             return
         active_optimizers = _get_active_optimizers(
-            self.trainer.optimizers, self.trainer.optimizer_frequencies, self.total_batch_idx
+            self.trainer.optimizers, self.trainer.optimizer_frequencies, self.batch_idx
```
Comment (tchaton) [inline src/pytorch_lightning/loops/epoch/training_epoch_loop.py:394]: Could you add a test to verify this works properly ?

Comment (milesial) [reply src/pytorch_lightning/loops/epoch/training_epoch_loop.py:394]: I modified the third case of `test_step_scheduling_for_multiple_optimizers_with_frequency` so that it tests that

File: setup.cfg [hunk 3]
```diff
@@ -34,6 +34,7 @@ markers =
     cloud:Run the cloud tests for example
 filterwarnings =
     error::FutureWarning
+    error:Detected call of `lr_scheduler.step\(\)` before `optimizer.step\(\)`:UserWarning
```
Comment (carmocca) [inline setup.cfg:37]: I added this line so that our CI fails if this warning appears. This way it tests that your patch works as expected.

**Review by carmocca — COMMENTED** [review]
Can you check the failing tests?

File: setup.cfg (hunk 3 above)
Comment (milesial) [reply setup.cfg:37]: Thanks, but this also makes IPU tests fail, this PR is focused on GPU. Not sure where to fix the issue on IPUs

**Comment by milesial** [comment]
I also modified `native_amp.py` in both lightning_fabric and pytorch_lightning. It doesn't seem like the lightning_fabric one is called for a typical workflow, so I don't know if it's the right way

File: setup.cfg (hunk 3 above)
Comment (akihironitta) [reply setup.cfg:37]: @milesial IMHO, we could just skip the IPU test failures as done in my old PR #11755 for now.

**Review by Borda — APPROVED** [review]

**Comment by mrembalski** [comment]
Hi @Borda, I also encounter the same issue. Will this be merged?
